simulation/halsim_ws_core/src/main/native/cpp/WSProvider_Analog.cpp

```diff
@@ -0,0 +1,112 @@
+      },                                                                   \
+      this, true)
+
+#define REGISTER_AOUT(halsim, jsonid, ctype, haltype)                       \
+  HALSIM_RegisterAnalogOut##halsim##Callback(                               \
+      m_channel,                                                            \
+      [](const char* name, void* param, const struct HAL_Value* value) {    \
+        static_cast<HALSimWSProviderAnalogOut*>(param)->ProcessHalCallback( \
+            {{jsonid, static_cast<ctype>(value->data.v_##haltype)}});       \
+      },                                                                    \
+      this, true)
+
+namespace wpilibws {
+
+void HALSimWSProviderAnalogIn::Initialize(WSRegisterFunc webRegisterFunc) {
+  CreateProviders<HALSimWSProviderAnalogIn>("AI", HAL_GetNumAnalogInputs(),
+                                            webRegisterFunc);
+}
+
+void HALSimWSProviderAnalogIn::RegisterCallbacks() {
+  m_initCbKey = REGISTER_AIN(Initialized, "<init", bool, boolean);
+  m_avgbitsCbKey = REGISTER_AIN(AverageBits, "<avg_bits", int32_t, int);
+  m_oversampleCbKey =
+      REGISTER_AIN(OversampleBits, "<oversample_bits", int32_t, int);
+  m_voltageCbKey = REGISTER_AIN(Voltage, ">voltage", double, double);
+
+  m_accumInitCbKey = REGISTER_AIN_ACCUM(Initialized, "<init", bool, boolean);
+  m_accumValueCbKey = REGISTER_AIN_ACCUM(Value, ">value", int64_t,
+                                         long);  // NOLINT(runtime/int)
+  m_accumCountCbKey = REGISTER_AIN_ACCUM(Count, ">count", int64_t,
+                                         long);  // NOLINT(runtime/int)
+  m_accumCenterCbKey = REGISTER_AIN_ACCUM(Center, "<center", int32_t, int);
+  m_accumDeadbandCbKey =
+      REGISTER_AIN_ACCUM(Deadband, "<deadband", int32_t, int);
+}
+
+void HALSimWSProviderAnalogIn::CancelCallbacks() {
+  // Cancel callbacks
+  HALSIM_CancelAnalogInInitializedCallback(m_channel, m_initCbKey);
+  HALSIM_CancelAnalogInAverageBitsCallback(m_channel, m_avgbitsCbKey);
+  HALSIM_CancelAnalogInOversampleBitsCallback(m_channel, m_oversampleCbKey);
+  HALSIM_CancelAnalogInVoltageCallback(m_channel, m_voltageCbKey);
+  HALSIM_CancelAnalogInAccumulatorInitializedCallback(m_channel,
+                                                      m_accumInitCbKey);
+  HALSIM_CancelAnalogInAccumulatorValueCallback(m_channel, m_accumValueCbKey);
+  HALSIM_CancelAnalogInAccumulatorCountCallback(m_channel, m_accumCountCbKey);
+  HALSIM_CancelAnalogInAccumulatorCenterCallback(m_channel, m_accumCenterCbKey);
+  HALSIM_CancelAnalogInAccumulatorDeadbandCallback(m_channel,
+                                                   m_accumDeadbandCbKey);
+}
+
+void HALSimWSProviderAnalogIn::OnNetValueChanged(const wpi::json& json) {
+  wpi::json::const_iterator it;
+  if ((it = json.find(">voltage")) != json.end()) {
+    HALSIM_SetAnalogInVoltage(m_channel, it.value());
+  }
+  if ((it = json.find("accum")) != json.end()) {
```

Comment:
Ok, you're going to hate me now, but bear with me here.

Since this whole thing has been switched to type/value and a flat namespace, *and* joysticks now are completely flat (I think?), this is the only remaining nested json space.

I think it should just be flattened for consistency? `<accum_init`, ... etc

Sorry. 😿

Reply:
This was marked as resolved, but only the send has been fixed, OnNetValueChanged side hasn't been fixed?